File: service_frontend/lib/tenant.ts
/**
 * Tenant slug rules + resolution (plan 07 §4, §6).
 *
 * The slug is the tenant's URL identity (subdomain) — lowercase kebab, immutable
 * after creation. `deriveTenantSlug` resolves the current tenant from the
 * hostname (subdomain-based resolution, D4); local dev falls back to
 * NEXT_PUBLIC_TENANT_SLUG, then "default".
 */

/** Slugs that can never be tenant identities (infra/system hostnames). */
export const RESERVED_TENANT_SLUGS: readonly string[] = [
  'www',
  'api',
  'app',
  'admin',
  'platform',
  'default',
  'mail',
  'ftp',
  'assets',
  'static',
  'docs',
  'status',
  'support',
  'billing',
];

/** lowercase kebab, 3–63 chars, no leading/trailing/double hyphen. */
const SLUG_RE = /^[a-z0-9](?:-?[a-z0-9]){2,62}$/;

export function isValidTenantSlug(slug: string): boolean {
  return SLUG_RE.test(slug);
}

export function isReservedTenantSlug(slug: string): boolean {
  return RESERVED_TENANT_SLUGS.includes(slug);
}

/** Hosts that can never carry a tenant subdomain (local dev / raw IPs). */
const NON_TENANT_HOSTS = new Set(['localhost', '127.0.0.1', '0.0.0.0']);

/**
 * Subdomains that are infra aliases, never tenant homes. This is the reserved
 * list MINUS `platform`/`default` — those are real (seeded) tenants and must
 * resolve (the operator console lives at platform.<domain>).
 */
const NON_TENANT_SUBDOMAINS = new Set(
  RESERVED_TENANT_SLUGS.filter((s) => s !== 'platform' && s !== 'default'),
);

const FALLBACK_SLUG = process.env.NEXT_PUBLIC_TENANT_SLUG || 'default';

/**
 * Build a tenant's URL from the host the operator is browsing on —
 * swap the subdomain for the tenant slug, keep domain/port/protocol.
 * platform.localhost:3001 + acme → http://acme.localhost:3001;
 * platform.foundryxems.com + acme → https://acme.foundryxems.com.
 */
export function tenantUrl(slug: string, location: Location): string {
  const [host, port] = location.host.toLowerCase().split(':');
  const labels = host.split('.');
  // Base domain = everything after the current subdomain; bare hosts
  // (localhost, foundryxems.com) are already the base.
  const isSubdomained = labels.length >= 3 || (labels.length === 2 && labels[1] === 'localhost');
  const base = isSubdomained ? labels.slice(1).join('.') : host;
  return `${location.protocol}//${slug}.${base}${port ? `:${port}` : ''}`;
}

/**
 * Resolve the tenant slug from a hostname: first label of a 3+-label host
 * (`acme.foundryxems.com` → `acme`) or of `<slug>.localhost` (dev), unless it's
 * an infra alias (`www`, `api`, …). Bare domains, localhost and IPs fall back
 * to the dev/default slug.
 */
export function deriveTenantSlug(hostname: string): string {
  const host = hostname.toLowerCase().split(':')[0];
  if (NON_TENANT_HOSTS.has(host) || /^\d+(\.\d+){3}$/.test(host)) return FALLBACK_SLUG;

  const labels = host.split('.');
  // `<slug>.localhost` (browsers resolve *.localhost to 127.0.0.1) or a
  // subdomain + domain + tld host (acme.foundryxems.com).
  const isDevSubdomain = labels.length === 2 && labels[1] === 'localhost';
  if (!isDevSubdomain && labels.length < 3) return FALLBACK_SLUG;

  const candidate = labels[0];
  if (NON_TENANT_SUBDOMAINS.has(candidate) || !isValidTenantSlug(candidate)) return FALLBACK_SLUG;
  return candidate;
}

File: dreamz_ems_backend/app/models/tenant.py
"""Tenant model — root of multi-tenancy (SaaS, plan 07).

A tenant is a workspace: users/roles/data are scoped to it via ``tenant_id``.
Lifecycle is a core ``statuses`` row (FK) — behavior binds to the status TRAIT
FLAGS (``blocks_access`` / ``is_archived``, sprint-2/01 D2); transitions move
along the status-engine edge graph only. Labels/colors stay editable.

The reserved **platform tenant** (``is_platform``) hosts the operator team;
platform endpoints require membership in it (``require_platform_permission``).
"""
import re
import uuid

from sqlalchemy import Boolean, Column, ForeignKey, String
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.models.utc_datetime import UTCDateTime

from app.database import Base

# Stable id/slug for the bootstrap tenant every row falls back to until tenant
# resolution is explicit everywhere. Keep in sync with scripts/init_db.py.
DEFAULT_TENANT_ID = "00000000-0000-0000-0000-000000000001"
DEFAULT_TENANT_SLUG = "default"
DEFAULT_TENANT_NAME = "Dreamz EMS"

# The reserved operator tenant (plan 07 §5) — seeded at bootstrap.
PLATFORM_TENANT_ID = "00000000-0000-0000-0000-000000000002"
PLATFORM_TENANT_SLUG = "platform"
PLATFORM_TENANT_NAME = "Dreamz Platform"

# Slugs that can never be tenant identities (infra/system hostnames). Mirrors
# the frontend list in lib/tenant.ts.
RESERVED_TENANT_SLUGS = frozenset(
    {
        "www",
        "api",
        "app",
        "admin",
        "platform",
        "default",
        "mail",
        "ftp",
        "assets",
        "static",
        "docs",
        "status",
        "support",
        "billing",
    }
)

# lowercase kebab, 3-63 chars, no leading/trailing/double hyphen.
_SLUG_RE = re.compile(r"^[a-z0-9](?:-?[a-z0-9]){2,62}$")


def is_valid_tenant_slug(slug: str) -> bool:
    return bool(_SLUG_RE.match(slug))


class Tenant(Base):
    __tablename__ = "tenants"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    name = Column(String, nullable=False)
    # URL identity (subdomain) — immutable after creation.
    slug = Column(String, unique=True, nullable=False, index=True)
    # Lifecycle — a status-engine row; behavior binds to its trait flags.
    status_id = Column(String, ForeignKey("statuses.id"), nullable=False, index=True)
    # Exactly one seeded row true — the operator tenant.
    is_platform = Column(Boolean, nullable=False, default=False)
    contact_name = Column(String, nullable=True)
    contact_email = Column(String, nullable=True)
    # Schema-ready; CNAME/infra wiring is BL-034.
    custom_domain = Column(String, unique=True, nullable=True)
    notes = Column(String, nullable=True)

    created_at = Column(
        UTCDateTime(), server_default=func.now(), nullable=False
    )
    updated_at = Column(
        UTCDateTime(),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    # joined: the lifecycle check in get_current_user/login reads the category
    # on every request — load it with the tenant row, never lazily/N+1.
    status = relationship("Status", lazy="joined")

    @property
    def signin_allowed(self) -> bool:
        """Single chokepoint for the lifecycle rule (login + per-request).

        Behavior binds to the status TRAIT FLAGS (sprint-2/01 D2) — a status
        that blocks access or archives the tenant kills sign-in; a missing
        status row counts as not-allowed (defensive — FK should prevent it).
        """
        return (
            self.status is not None
            and not self.status.blocks_access
            and not self.status.is_archived
        )
